src/herder/HerderImpl.h

```diff
@@ -288,6 +289,10 @@ class HerderImpl : public Herder
     Application& mApp;
     LedgerManager& mLedgerManager;
 
+    // Set of nodes whose SCP messages from the previous ledger have already
+    // been stored in the scphistory database table.
+    UnorderedSet<NodeID> mPrevExternalizedEnvs;
```

Comment:
This isn't _quite_ right: current implementation only records new statements from nodes we haven't heard from at all in the last consensus round. The scenario we also want to handle is when we receive _newer_ SCP messages from nodes in our quorum set. This can occur if, for example, we have CONFIRM messages from our quorum slice (which is enough to externalize that ledger), but then receive EXTERNALIZE messages from the same nodes after some time. We'd like to store those newer EXTERNALIZE messages as well, to make things easier for downstream systems (e.g. bridges). So I think we'd either want to store the diff in SCP messages, or rewrite the whole externalizing state, depending on the performance impact.

Reply:
Ah, this was an oversight on my part, I forgot about that scenario. I'll fix this and do another round of performance testing.

Reply:
I pushed up a fix for this in 07defe8ceb992f810e47f5047ae60fd7dfb58dfe.  Weirdly enough the performance of rewriting the entire externalizing state is faster than tracking diffs. I think this is because rewriting can be done with a single `DELETE` followed by a single bulk `INSERT` (I changed `saveSCPHistory` to perform a single multi-row `INSERT` into `scphistory`, which provided a small performance improvement), while tracking diffs requires multiple `MODIFY`s and `INSERT`s.

Regardless, the performance impact of all of this isn't too bad. Compared to `master` on pubnet, the new process of saving SCP messages (doing two `saveSCPHistory` calls) takes ~3.7ms longer on Postgres and ~0.9ms longer on SQLite on my machine. In relative terms, that's 31% longer for Postgres and 50% longer for SQLite.